src/GitHub.App/Services/TeamExplorerContext.cs

```diff
@@ -43,28 +47,56 @@ public class TeamExplorerContext : ITeamExplorerContext
         ILocalRepositoryModel repositoryModel;
 
         [ImportingConstructor]
+        TeamExplorerContext(
+            IVSGitExt gitExt,
+            [Import(typeof(SVsServiceProvider))] IServiceProvider sp,
+            IPullRequestService pullRequestService) : this(
+                gitExt,
+                new AsyncLazy<DTE>(async () =>
+                {
+                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    return (DTE)sp.GetService(typeof(DTE));
+                }),
+                pullRequestService,
+                ThreadHelper.JoinableTaskContext)
+        {
+        }
+
         public TeamExplorerContext(
-            IGitHubServiceProvider serviceProvider,
             IVSGitExt gitExt,
-            IPullRequestService pullRequestService)
+            AsyncLazy<DTE> dteAsync,
+            IPullRequestService pullRequestService,
+            JoinableTaskContext joinableTaskContext)
         {
+            JoinableTaskCollection = joinableTaskContext.CreateCollection();
+            JoinableTaskCollection.DisplayName = nameof(TeamExplorerContext);
+            JoinableTaskFactory = joinableTaskContext.CreateFactory(JoinableTaskCollection);
+            
             this.gitExt = gitExt;
+            this.dteAsync = dteAsync;
             this.pullRequestService = pullRequestService;
 
-            // This is a standard service which should always be available.
-            dte = serviceProvider.GetService<DTE>();
-
-            Refresh();
+            StartRefresh();
             gitExt.ActiveRepositoriesChanged += Refresh;
         }
 
-        async void Refresh()
+        void StartRefresh() => JoinableTaskFactory.Context.Factory.RunAsync(QueueRefreshAsync).Task.FileAndForget(log);
+        void Refresh() => JoinableTaskFactory.Context.Factory.Run(QueueRefreshAsync);
+
+        async Task QueueRefreshAsync()
+        {
+            await JoinableTaskCollection.JoinTillEmptyAsync();
+            await JoinableTaskFactory.RunAsync(RefreshAsync);
```

Comment:
The most important thing is that the data from the most recent call refresh is up to date. I know that just becuase task A starts befoire task B, doesn't mean that it will complete before task B. If task A happened to complete after task B, it might set `ActiveRepository` back to null (which has been a problem in the past).

Since `RefreshAsync` is relatively quick and I'm going to use Rx for throttling, I'm inclined to keep it simple and just queue up refreshes. I could maybe use a `CancellationToken`, but I think this would be optimzing an edge case.